cpp/tensorrt_llm/kernels/unfusedAttentionKernels/unfusedAttentionKernels_2_template.h

```diff
@@ -640,21 +639,32 @@ __global__ void applyBiasRopeUpdateKVCache(QKVPreprocessingParams<T, KVCacheBuff
             params.fmha_tile_counter[0] = 0u;
         }
         // Take the quantization scales into consideration.
-        float q_scale_quant_orig = params.q_scale_quant_orig ? params.q_scale_quant_orig[0] : 1.f;
-        float kv_scale_quant_orig = params.kv_scale_quant_orig ? params.kv_scale_quant_orig[0] : 1.f;
+        float q_scale_quant_orig, k_scale_quant_orig, v_scale_quant_orig;
+        if constexpr (ENABLE_4BITS_CACHE)
+        {
+            q_scale_quant_orig = params.qkv_scale_quant_orig ? params.qkv_scale_quant_orig[0] : 1.f;
+            k_scale_quant_orig = params.qkv_scale_quant_orig ? params.qkv_scale_quant_orig[1] : 1.f;
+            v_scale_quant_orig = params.qkv_scale_quant_orig ? params.qkv_scale_quant_orig[2] : 1.f;
+        }
+        else
+        {
+            q_scale_quant_orig = params.qkv_scale_quant_orig ? params.qkv_scale_quant_orig[0] : 1.f;
```

Comment:
why not we always use three separate scales even though they might be the same for FP8 kv cache.

Reply:
It is for compatibility -- we did not consider trt backend, and many other places (like MMHA) still reads from `kv_scale_quant_orig[0]`. We can do a further cleanup for this purpose, what do you think?

Reply:
okay. we indeed need to clean a lot of things (especially when trt backend can be removed one day). Thanks again!